ui/src/treeDataProvider/AzExtParentTreeItem.ts

```diff
@@ -0,0 +1,324 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as path from 'path';
+import { commands, TreeItemCollapsibleState } from 'vscode';
+import * as types from '../../index';
+import { NotImplementedError } from '../errors';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import { AzExtTreeItem } from './AzExtTreeItem';
+import { GenericTreeItem } from './GenericTreeItem';
+import { IAzExtParentTreeItemInternal, isAzExtParentTreeItem } from './InternalInterfaces';
+import { loadingIconPath, loadMoreLabel } from './treeConstants';
+
+export abstract class AzExtParentTreeItem extends AzExtTreeItem implements types.AzExtParentTreeItem, IAzExtParentTreeItemInternal {
+    //#region Properties implemented by base class
+    public childTypeLabel?: string;
+    public autoSelectInTreeItemPicker?: boolean;
+    //#endregion
+
+    public readonly collapsibleState: TreeItemCollapsibleState | undefined = TreeItemCollapsibleState.Collapsed;
+    public readonly _isAzExtParentTreeItem: boolean = true;
+
+    private _cachedChildren: AzExtTreeItem[] = [];
+    private _creatingTreeItems: AzExtTreeItem[] = [];
+    private _clearCache: boolean = true;
+    private _loadMoreChildrenTask: Promise<void> | undefined;
+    private _initChildrenTask: Promise<void> | undefined;
+
+    public async getCachedChildren(): Promise<AzExtTreeItem[]> {
+        if (this._clearCache) {
+            this._initChildrenTask = this.loadMoreChildren();
+        }
+
+        if (this._initChildrenTask) {
+            await this._initChildrenTask;
+        }
+
+        return this._cachedChildren;
+    }
+
+    public get creatingTreeItems(): AzExtTreeItem[] {
+        return this._creatingTreeItems;
+    }
+
+    //#region Methods implemented by base class
+    public abstract loadMoreChildrenImpl(clearCache: boolean): Promise<AzExtTreeItem[]>;
+    public abstract hasMoreChildrenImpl(): boolean;
+    // tslint:disable-next-line:no-any
+    public createChildImpl?(showCreatingTreeItem: (label: string) => void, userOptions?: any): Promise<AzExtTreeItem>;
```

Comment:
Can we switch to unknown?

Reply:
That would require updating the version of typescript to at least 3.0.0. I can do that in a separate PR